src/ProductConstructionService/ProductConstructionService.Api/Queue/JobsProcessor.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Azure.Storage.Queues;
+using Azure.Storage.Queues.Models;
+using Microsoft.Extensions.Options;
+using ProductConstructionService.Api.Queue.Jobs;
+
+namespace ProductConstructionService.Api.Queue;
+
+public class JobsProcessor(
+    ILogger<JobsProcessor> logger,
+    IOptions<JobProcessorOptions> options,
+    JobsProcessorScopeManager scopeManager,
+    QueueServiceClient queueServiceClient)
+    : BackgroundService
+{
+    private readonly ILogger<JobsProcessor> _logger = logger;
+    private readonly IOptions<JobProcessorOptions> _options = options;
+    private readonly JobsProcessorScopeManager _scopeManager = scopeManager;
+
+    protected override async Task ExecuteAsync(CancellationToken cancellationToken)
+    {
+        // The API won't be initialized until the BackgroundService goes async. Since the scopeManagers blocks aren't async, we have to do it here
+        await Task.Delay(1000);
+
+        QueueClient queueClient = queueServiceClient.GetQueueClient(_options.Value.JobQueueName);
+        _logger.LogInformation("Starting to process PCS jobs {queueName}", _options.Value.JobQueueName);
+        while (!cancellationToken.IsCancellationRequested)
+        {
+            using (_scopeManager.BeginJobScopeWhenReady(cancellationToken))
+            {
+                try
+                {
+                    await ReadAndProcessJobAsync(queueClient, cancellationToken);
+                }
+                // If the cancellation token gets cancelled, we just want to exit
+                catch (OperationCanceledException)
+                {
+                    break;
+                }
```

Comment:
So.. this quite hasn't solved the problem I was commenting about and actually added one more:
- We still don't catch the `OperationCanceledException` from `BeginJobScopeWhenReady`
- Now we don't long an exception if we get cancellation mid-job and we will never know where it got cancelled and that it got cancelled

Reply:
Ah yes, I missed that. Maybe we shouldn't throw from `BeginJobScopeWhenReady`, but rather check the cancelation token when we're in the scope and just break out of the while loop? The part I don't like there is that we'd be entering the scope in the first place tho

Reply:
Yes, that might be better

Reply:
I moved the cancellationToken check outside of the scopeManager